File: sandbox/cybersecurity/cryptography/HistoricalCiphersDemo/Ciphers/Vigenère/VigenèreCipher.cs
namespace HistoricalCiphersDemo.Ciphers.Vigenère;

public class VigenèreCipher: ICipher
{
    public string Name => "Vigenère";
    
    public string Description => "A polyalphabetic substitution cipher that uses a keyword to determine shifting.";

    public string Encrypt(string text, object key)
    {
        return ProcessText(text, key.ToString()!, true);
    }   

    public string Decrypt(string text, object key)
    {
        return ProcessText(text, key.ToString()!, false);
    }
    
    private string ProcessText(string text, string key, bool encrypt)
    {
        if (string.IsNullOrEmpty(text) || string.IsNullOrEmpty(key))
            return string.Empty;
            
        char[] result = new char[text.Length];
        int keyIndex = 0;
        
        for (int i = 0; i < text.Length; i++)
        {
            char textChar = text[i];
            
            // Skip non-alphabetic characters
            if (!char.IsLetter(textChar))
            {
                result[i] = textChar;
                continue;
            }
            
            // Get the corresponding key character (cycling through the key)
            // Only advance key index for alphabetic characters
            char keyChar = key[keyIndex % key.Length];
            keyIndex++;
            
            // Calculate the shift based on the key character
            int shift = char.ToUpper(keyChar) - 'A';
            
            // Handle encryption vs decryption
            if (!encrypt)
                shift = -shift;
                
            // Determine base value ('A' for uppercase, 'a' for lowercase)
            char baseChar = char.IsUpper(textChar) ? 'A' : 'a';
            
            // Apply the shift, ensuring positive modulo
            result[i] = (char)(((textChar - baseChar + shift + 26) % 26) + baseChar);
        }
        
        return new string(result);
    }
}

File: Projects/Testtrack/v240turbo - Historical Ciphers/Vigenère Cipher/vigenère_cipher.py
class VigenereCipher:
    @property
    def name(self):
        return "Vigenère"

    @property
    def description(self):
        return "A polyalphabetic substitution cipher that uses a keyword to determine shifting."

    def encrypt(self, text, key):
        return self._process_text(text, str(key), True)

    def decrypt(self, text, key):
        return self._process_text(text, str(key), False)

    def _process_text(self, text, key, encrypt):
        if not text or not key:
            return ""

        result = []
        key_index = 0

        for i in range(len(text)):
            text_char = text[i]

            if not text_char.isalpha():
                result.append(text_char)
                continue

            key_char = key[key_index % len(key)]
            key_index += 1

            shift = ord(key_char.upper()) - ord('A')

            if not encrypt:
                shift = -shift

            base_char = ord('A') if text_char.isupper() else ord('a')

            shifted_char = chr(
                ((ord(text_char) - base_char + shift + 26) % 26) + base_char)
            result.append(shifted_char)

        return ''.join(result)


if __name__ == "__main__":
    cipher = VigenereCipher()

    text = "Hello World!"
    key = "KEY"

    encrypted = cipher.encrypt(text, key)
    decrypted = cipher.decrypt(encrypted, key)

    print(f"Original: {text}")
    print(f"Key: {key}")
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")

    print("\n=== Another Example ===")
    text2 = "ATTACKATDAWN"
    key2 = "LEMON"

    encrypted2 = cipher.encrypt(text2, key2)
    decrypted2 = cipher.decrypt(encrypted2, key2)

    print(f"Original: {text2}")
    print(f"Key: {key2}")
    print(f"Encrypted: {encrypted2}")
    print(f"Decrypted: {decrypted2}")
